### PR TITLE
Refactor autocomplete to query the server

### PR DESCRIPTION
**Pull request by jwilander**
#### Summary
This is a work in progress.

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-4266

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)
- [x] All new/modified APIs include changes to the drivers

### CODE REVIEW

**Comment by coreyhulen** [comment]
+1

File: model/client.go [hunk 1]
```diff
@@ -584,6 +584,21 @@ func (c *Client) SearchUsers(term string, teamId string, options map[string]stri
 	}
 }
 
+// AutocompleteUsers returns two lists for autocompletion of user mentions. The first list "in",
+// specifies users in the current context (the channel, for example). The second list "out" specifies
+// users outside of that context (users not in the channel, for example). Username is required, channel
+// id is optional. Must be authenticated.
+func (c *Client) AutocompleteUsers(username string, channelId string) (*Result, *AppError) {
```
Comment (hmhealey) [inline model/client.go]: Maybe change this to return the `UserAutocomplete` object? I think we were going to move away from just returning `Result` objects

File: webapp/actions/user_actions.jsx [hunk 2]
```diff
@@ -277,3 +277,20 @@ export function searchUsers(term, teamId = TeamStore.getCurrentId(), options = {
         }
     );
 }
+
+export function autocompleteUsers(username, channelId = '', success, error) {
```
Comment (hmhealey) [inline webapp/actions/user_actions.jsx]: Does this need the additional error handling that all the calls in `AsyncClient` have? Also, I don't think having a default parameter works if it's not the last parameter

File: webapp/components/suggestion/at_mention_provider.jsx [hunk 3]
```diff
@@ -99,31 +100,7 @@ class AtMentionSuggestion extends Suggestion {
...
-            (user.last_name && user.last_name.toLowerCase().startsWith(prefix)) ||
-            (user.nickname && user.nickname.toLowerCase().startsWith(prefix))) {
-            // create a new object here since we're mutating it by adding the type field
-            filtered.push(Object.assign({}, user, {type}));
-        }
-    }
-
-    return filtered;
-}
+let timeoutId = '';
```
Comment (hmhealey) [inline webapp/components/suggestion/at_mention_provider.jsx]: Can you make `timeoutId` a field of the `AtMentionProvider`? I know it's unlikely that we'll have multiple `AtMentionProvider`s querying the server at once, but it makes me leery having it not encapsulated in the object itself.

Comment (jwilander) [inline webapp/components/suggestion/at_mention_provider.jsx]: Yup, good call

File: model/client.go (hunk 1 above)
Comment (jwilander) [inline model/client.go]: Do any of the other driver functions do that? I'd rather be consistent and we can worry about switching them all over later

**Comment by jwilander** [comment]
It's not hard to add the ability to search by more than username but I'm not going to do that in this PR :P I'm just going to get us back to what we already have

File: model/client.go (hunk 1 above)
Comment (hmhealey) [inline model/client.go]: Most of the newer ones work that way, but we can just switch this one when we get around to switching all of the old ones

**Comment by hmhealey** [comment]
That's how it works on master if we're concerned with matching that :)

**Comment by jwilander** [comment]
Oh didn't realize we were doing that already, I might add it then

**Comment by jwilander** [comment]
Closing to resubmit

**Pull request closed without merging**